File: DynamicPrograming/php/nonAdjacentSum.php
<?php

declare(strict_types=1);

/** @param int[] $nums */
function nonAdjacentSum(array $nums): int
{
    $memo = [];
    $length = count($nums);

    $dfs = function (int $i) use (&$dfs, &$memo, $nums, $length): int {
        if ($i >= $length) {
            return 0;
        }
        if (array_key_exists($i, $memo)) {
            return $memo[$i];
        }

        $include = $nums[$i] + $dfs($i + 2);
        $exclude = $dfs($i + 1);
        return $memo[$i] = max($include, $exclude);
    };

    return $dfs(0);
}

File: DynamicPrograming/python/nonAdjacentSum.py
from functools import lru_cache
from typing import List


def nonAdjacentSum(nums: List[int]) -> int:
    @lru_cache(maxsize=None)
    def dfs(i: int) -> int:
        if i >= len(nums):
            return 0
        include = nums[i] + dfs(i + 2)
        exclude = dfs(i + 1)
        return max(include, exclude)

    return dfs(0)
